File: apps/web/src/lib/api/recommendations.ts
import type {
  AppliedRestaurantSearchFilters,
  RestaurantSearchFilters,
} from "./restaurants.js";

export const RECOMMENDATION_CATEGORIES = [
  "best_overall",
  "best_walkable",
  "best_public_transportation",
  "best_value",
  "most_consistently_recommended",
  "best_vegetarian_match",
  "hidden_gem",
  "mixed_reviews",
] as const;

export type RecommendationCategory =
  (typeof RECOMMENDATION_CATEGORIES)[number];

export type RecommendationBadge = {
  category: RecommendationCategory;
  label: string;
  reason: string;
};

export type RecommendationTheme = {
  theme: string;
  label: string;
  mention_count: number;
};

export type RecommendationTravel = {
  starting_area: string;
  walking_minutes: number;
  public_transit_minutes: number;
  driving_minutes: number;
  category:
    | "walkable"
    | "comfortable_walk"
    | "easy_public_transit"
    | "longer_public_transit"
    | "easy_drive"
    | "inconvenient";
  selected_mode: "walking" | "public_transit" | "driving";
  selected_minutes: number;
  explanation: string;
};

export type LatestInspection = {
  inspection_id: string;
  inspection_date: string;
  result: "Pass" | "Pass w/ Conditions";
  inspection_type: string;
  risk: string;
};

export type RankingFactor = {
  status: "active" | "not_applicable" | "unavailable";
  score: number | null;
  configured_weight: number;
  effective_weight: number;
  contribution: number;
  explanation: string;
};

export type RestaurantRecommendation = {
  rank: number;
  restaurant_id: string;
  name: string;
  address: string;
  latitude: number;
  longitude: number;
  cuisine: string;
  neighborhood: string;
  price_category: "$" | "$$" | "$$$";
  estimated_cost_per_person: number;
  vegetarian_available: boolean;
  vegan_available: boolean;
  rating: number;
  dataset_review_count: number;
  total_score: number;
  categories: RecommendationBadge[];
  travel: RecommendationTravel | null;
  top_positive_themes: RecommendationTheme[];
  top_negative_themes: RecommendationTheme[];
  review_confidence_score: number;
  review_confidence_band: "low" | "medium" | "high";
  ranking_factors: Record<string, RankingFactor>;
  ranking_explanation: string;
  latest_review_date: string;
  data_freshness_label: string;
  latest_inspection: LatestInspection;
};

export type RestaurantRecommendationResponse = {
  filters: AppliedRestaurantSearchFilters;
  configured_weights: Record<string, number>;
  match_count: number;
  assigned_categories: RecommendationCategory[];
  recommendations: RestaurantRecommendation[];
  data_notice: string;
};

const REQUEST_TIMEOUT_MS = 8_000;

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function isNullableString(value: unknown): value is string | null {
  return value === null || typeof value === "string";
}

function isNullableNumber(value: unknown): value is number | null {
  return value === null || typeof value === "number";
}

function isFilters(value: unknown): value is AppliedRestaurantSearchFilters {
  return (
    isRecord(value) &&
    isNullableString(value.cuisine) &&
    isNullableNumber(value.maximum_budget) &&
    typeof value.vegetarian_required === "boolean" &&
    isNullableString(value.starting_area) &&
    isNullableNumber(value.maximum_travel_time)
  );
}

function isBadge(value: unknown): value is RecommendationBadge {
  return (
    isRecord(value) &&
    RECOMMENDATION_CATEGORIES.includes(
      value.category as RecommendationCategory,
    ) &&
    typeof value.label === "string" &&
    typeof value.reason === "string"
  );
}

function isTheme(value: unknown): value is RecommendationTheme {
  return (
    isRecord(value) &&
    typeof value.theme === "string" &&
    typeof value.label === "string" &&
    typeof value.mention_count === "number"
  );
}

function isTravel(value: unknown): value is RecommendationTravel {
  return (
    isRecord(value) &&
    typeof value.starting_area === "string" &&
    typeof value.walking_minutes === "number" &&
    typeof value.public_transit_minutes === "number" &&
    typeof value.driving_minutes === "number" &&
    typeof value.category === "string" &&
    (value.selected_mode === "walking" ||
      value.selected_mode === "public_transit" ||
      value.selected_mode === "driving") &&
    typeof value.selected_minutes === "number" &&
    typeof value.explanation === "string"
  );
}

function isLatestInspection(value: unknown): value is LatestInspection {
  return (
    isRecord(value) &&
    typeof value.inspection_id === "string" &&
    typeof value.inspection_date === "string" &&
    (value.result === "Pass" || value.result === "Pass w/ Conditions") &&
    typeof value.inspection_type === "string" &&
    typeof value.risk === "string"
  );
}

function isFactor(value: unknown): value is RankingFactor {
  return (
    isRecord(value) &&
    (value.status === "active" ||
      value.status === "not_applicable" ||
      value.status === "unavailable") &&
    isNullableNumber(value.score) &&
    typeof value.configured_weight === "number" &&
    typeof value.effective_weight === "number" &&
    typeof value.contribution === "number" &&
    typeof value.explanation === "string"
  );
}

function isRecommendation(value: unknown): value is RestaurantRecommendation {
  if (!isRecord(value) || !isRecord(value.ranking_factors)) {
    return false;
  }
  return (
    typeof value.rank === "number" &&
    typeof value.restaurant_id === "string" &&
    typeof value.name === "string" &&
    typeof value.address === "string" &&
    typeof value.latitude === "number" &&
    typeof value.longitude === "number" &&
    typeof value.cuisine === "string" &&
    typeof value.neighborhood === "string" &&
    (value.price_category === "$" ||
      value.price_category === "$$" ||
      value.price_category === "$$$") &&
    typeof value.estimated_cost_per_person === "number" &&
    typeof value.vegetarian_available === "boolean" &&
    typeof value.vegan_available === "boolean" &&
    typeof value.rating === "number" &&
    typeof value.dataset_review_count === "number" &&
    typeof value.total_score === "number" &&
    Array.isArray(value.categories) &&
    value.categories.every(isBadge) &&
    (value.travel === null || isTravel(value.travel)) &&
    Array.isArray(value.top_positive_themes) &&
    value.top_positive_themes.every(isTheme) &&
    Array.isArray(value.top_negative_themes) &&
    value.top_negative_themes.every(isTheme) &&
    typeof value.review_confidence_score === "number" &&
    (value.review_confidence_band === "low" ||
      value.review_confidence_band === "medium" ||
      value.review_confidence_band === "high") &&
    Object.values(value.ranking_factors).every(isFactor) &&
    typeof value.ranking_explanation === "string" &&
    typeof value.latest_review_date === "string" &&
    typeof value.data_freshness_label === "string" &&
    isLatestInspection(value.latest_inspection)
  );
}

export function parseRestaurantRecommendationResponse(
  value: unknown,
): RestaurantRecommendationResponse | null {
  if (
    !isRecord(value) ||
    !isFilters(value.filters) ||
    !isRecord(value.configured_weights) ||
    !Object.values(value.configured_weights).every(
      (weight) => typeof weight === "number",
    ) ||
    typeof value.match_count !== "number" ||
    !Array.isArray(value.assigned_categories) ||
    !value.assigned_categories.every((category) =>
      RECOMMENDATION_CATEGORIES.includes(category as RecommendationCategory),
    ) ||
    !Array.isArray(value.recommendations) ||
    !value.recommendations.every(isRecommendation) ||
    value.match_count !== value.recommendations.length ||
    typeof value.data_notice !== "string"
  ) {
    return null;
  }
  return value as RestaurantRecommendationResponse;
}

export function buildRecommendationRequest(
  filters: RestaurantSearchFilters,
): { filters: Record<string, string | number | boolean> } {
  const requestFilters: Record<string, string | number | boolean> = {
    vegetarian_required: filters.vegetarianRequired,
  };
  if (filters.cuisine) requestFilters.cuisine = filters.cuisine;
  if (filters.maximumBudget !== undefined) {
    requestFilters.maximum_budget = filters.maximumBudget;
  }
  if (filters.startingArea) requestFilters.starting_area = filters.startingArea;
  if (filters.maximumTravelTime !== undefined) {
    requestFilters.maximum_travel_time = filters.maximumTravelTime;
  }
  return { filters: requestFilters };
}

function readErrorMessage(value: unknown): string | null {
  if (!isRecord(value)) return null;
  if (typeof value.detail === "string") return value.detail;
  if (isRecord(value.detail) && typeof value.detail.message === "string") {
    return value.detail.message;
  }
  if (Array.isArray(value.detail) && isRecord(value.detail[0])) {
    const message = value.detail[0].msg;
    return typeof message === "string" ? message.replace(/^Value error, /, "") : null;
  }
  return null;
}

export async function recommendRestaurants(
  filters: RestaurantSearchFilters,
): Promise<RestaurantRecommendationResponse> {
  let response: Response;
  try {
    response = await fetch("/api/restaurants/recommendations", {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(buildRecommendationRequest(filters)),
      signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
    });
  } catch {
    throw new Error(
      "BiteCheck could not reach the recommendation service. Please try again.",
    );
  }

  let payload: unknown;
  try {
    payload = await response.json();
  } catch {
    throw new Error("BiteCheck received an unreadable response.");
  }
  if (!response.ok) {
    throw new Error(
      readErrorMessage(payload) ??
        `The recommendation request failed with status ${response.status}.`,
    );
  }
  const parsed = parseRestaurantRecommendationResponse(payload);
  if (parsed === null) {
    throw new Error("BiteCheck received an unexpected recommendation response.");
  }
  return parsed;
}

File: apps/api/src/bitecheck_api/restaurants/recommendations.py
from statistics import median

from bitecheck_api.restaurants.confidence import (
    ReviewConfidenceDataError,
    ReviewConfidenceRepository,
    confidence_scores,
)
from bitecheck_api.restaurants.insights import (
    RecommendationInsightsDataError,
    RecommendationInsightsRepository,
)
from bitecheck_api.restaurants.models import (
    RecommendationBadge,
    RecommendationCategory,
    RecommendationTravel,
    RestaurantRecommendationCard,
    RestaurantRecommendationResponse,
    RestaurantReviewConfidence,
    RestaurantRankingRequest,
    RankingWeights,
    TravelCategoryThresholds,
)
from bitecheck_api.restaurants.ranking import RestaurantRankingService
from bitecheck_api.restaurants.repository import RestaurantRepository
from bitecheck_api.restaurants.travel import categorize_travel


CATEGORY_ORDER: tuple[RecommendationCategory, ...] = (
    "best_overall",
    "best_walkable",
    "best_public_transportation",
    "best_value",
    "most_consistently_recommended",
    "best_vegetarian_match",
    "hidden_gem",
    "mixed_reviews",
)

CATEGORY_LABELS: dict[RecommendationCategory, str] = {
    "best_overall": "Best overall",
    "best_walkable": "Best walkable",
    "best_public_transportation": "Best by public transportation",
    "best_value": "Best value",
    "most_consistently_recommended": "Most consistently recommended",
    "best_vegetarian_match": "Best vegetarian match",
    "hidden_gem": "Hidden gem",
    "mixed_reviews": "Mixed reviews — proceed carefully",
}


class RestaurantRecommendationService:
    def __init__(
        self,
        restaurant_repository: RestaurantRepository,
        confidence_repository: ReviewConfidenceRepository,
        insights_repository: RecommendationInsightsRepository,
        default_weights: RankingWeights,
    ) -> None:
        self._restaurant_repository = restaurant_repository
        self._confidence_repository = confidence_repository
        self._insights_repository = insights_repository
        self._default_weights = default_weights

    def recommend(
        self, request: RestaurantRankingRequest
    ) -> RestaurantRecommendationResponse:
        ranking = RestaurantRankingService(
            self._restaurant_repository,
            self._confidence_repository,
            self._default_weights,
        ).rank(request)
        records = {
            item.restaurant_id: item
            for item in self._restaurant_repository.list_restaurants()
        }
        confidence = confidence_scores(self._confidence_repository)
        insights = {
            item.restaurant_id: item
            for item in self._insights_repository.get_all().restaurants
        }
        ranked_ids = {item.restaurant_id for item in ranking.rankings}
        if not ranked_ids.issubset(confidence):
            raise ReviewConfidenceDataError(
                "Review confidence does not cover every ranked restaurant."
            )
        if not ranked_ids.issubset(insights):
            raise RecommendationInsightsDataError(
                "Recommendation insights do not cover every ranked restaurant."
            )

        cards: list[RestaurantRecommendationCard] = []
        for ranked in ranking.rankings:
            record = records[ranked.restaurant_id]
            confidence_item = confidence[ranked.restaurant_id]
            insight = insights[ranked.restaurant_id]
            travel: RecommendationTravel | None = None
            if ranking.filters.starting_area is not None:
                estimate = record.estimated_transportation[
                    ranking.filters.starting_area
                ]
                thresholds = TravelCategoryThresholds(
                    maximum_acceptable_minutes=(
                        ranking.filters.maximum_travel_time or 50
                    )
                )
                decision = categorize_travel(estimate, thresholds)
                travel = RecommendationTravel(
                    starting_area=ranking.filters.starting_area,
                    walking_minutes=estimate.walking_minutes,
                    public_transit_minutes=estimate.public_transit_minutes,
                    driving_minutes=estimate.driving_minutes,
                    category=decision.category,
                    selected_mode=decision.selected_mode,
                    selected_minutes=decision.selected_minutes,
                    explanation=decision.explanation,
                )

            cards.append(
                RestaurantRecommendationCard(
                    rank=ranked.rank,
                    restaurant_id=ranked.restaurant_id,
                    name=record.name,
                    address=record.address,
                    latitude=record.latitude,
                    longitude=record.longitude,
                    cuisine=record.cuisine,
                    neighborhood=record.neighborhood,
                    price_category=record.price_category,
                    estimated_cost_per_person=record.estimated_cost_per_person,
                    vegetarian_available=record.vegetarian_available,
                    vegan_available=record.vegan_available,
                    rating=record.rating,
                    dataset_review_count=record.review_count,
                    total_score=ranked.total_score,
                    categories=[],
                    travel=travel,
                    top_positive_themes=insight.top_positive_themes,
                    top_negative_themes=insight.top_negative_themes,
                    review_confidence_score=(
                        confidence_item.review_confidence_score
                    ),
                    review_confidence_band=confidence_item.confidence_band,
                    ranking_factors=ranked.factors,
                    ranking_explanation=ranked.summary,
                    latest_review_date=insight.latest_review_date,
                    data_freshness_label=(
                        f"Latest synthetic review: {insight.latest_review_date}"
                    ),
                    latest_inspection=record.latest_inspection,
                )
            )

        assigned = self._assign_categories(
            cards,
            confidence,
            ranking.filters.maximum_travel_time or 50,
        )
        return RestaurantRecommendationResponse(
            filters=ranking.filters,
            configured_weights=ranking.configured_weights,
            match_count=len(cards),
            assigned_categories=assigned,
            recommendations=cards,
            data_notice=(
                "Restaurant identities, addresses, coordinates, and inspection "
                "records come from a fixed City of Chicago snapshot dated "
                "2026-08-30. Cuisines, prices, dietary flags, ratings, reviews, "
                "themes, confidence, hours, and travel times are synthetic demo data."
            ),
        )

    @staticmethod
    def _assign_categories(
        cards: list[RestaurantRecommendationCard],
        confidence: dict[str, RestaurantReviewConfidence],
        maximum_acceptable_minutes: int,
    ) -> list[RecommendationCategory]:
        if not cards:
            return []

        assignments: dict[RecommendationCategory, tuple[str, str]] = {}
        assignments["best_overall"] = (
            cards[0].restaurant_id,
            f"Highest weighted recommendation score at {cards[0].total_score:.2f}/100.",
        )

        walkable = [
            card
            for card in cards
            if card.travel is not None and card.travel.selected_mode == "walking"
        ]
        if walkable:
            winner = walkable[0]
            winner_travel = winner.travel
            if winner_travel is None:
                raise AssertionError("walkable candidates must include travel data")
            assignments["best_walkable"] = (
                winner.restaurant_id,
                f"Highest-ranked walking option at {winner_travel.selected_minutes} minutes.",
            )

        transit = [
            card
            for card in cards
            if card.travel is not None
            and card.travel.public_transit_minutes
            <= maximum_acceptable_minutes
        ]
        if transit:
            winner = transit[0]
            winner_travel = winner.travel
            if winner_travel is None:
                raise AssertionError("transit candidates must include travel data")
            assignments["best_public_transportation"] = (
                winner.restaurant_id,
                f"Highest-ranked transit option at {winner_travel.public_transit_minutes} minutes.",
            )

        value = min(
            cards,
            key=lambda card: (card.estimated_cost_per_person, -card.total_score),
        )
        assignments["best_value"] = (
            value.restaurant_id,
            f"Lowest estimated cost among these matches at ${value.estimated_cost_per_person} per person.",
        )

        consistent = max(
            cards,
            key=lambda card: (card.review_confidence_score, card.total_score),
        )
        assignments["most_consistently_recommended"] = (
            consistent.restaurant_id,
            f"Highest Review Confidence in this result set at {consistent.review_confidence_score:.2f}/100.",
        )

        vegetarian = [card for card in cards if card.vegetarian_available]
        if vegetarian:
            winner = vegetarian[0]
            assignments["best_vegetarian_match"] = (
                winner.restaurant_id,
                "Highest-ranked result with synthetic vegetarian availability.",
            )

        review_count_median = median(card.dataset_review_count for card in cards)
        hidden_gems = [
            card
            for card in cards
            if card.dataset_review_count <= review_count_median
            and card.rating >= 4
            and card.review_confidence_score >= 65
        ]
        if hidden_gems:
            winner = hidden_gems[0]
            assignments["hidden_gem"] = (
                winner.restaurant_id,
                f"Strong score with {winner.dataset_review_count} dataset ratings, below this result set's median.",
            )

        mixed = min(
            cards,
            key=lambda card: (
                confidence[card.restaurant_id]
                .components["rating_consistency"]
                .score,
                -sum(theme.mention_count for theme in card.top_negative_themes),
            ),
        )
        consistency_score = confidence[mixed.restaurant_id].components[
            "rating_consistency"
        ].score
        if consistency_score < 60:
            assignments["mixed_reviews"] = (
                mixed.restaurant_id,
                f"Rating-consistency component is {consistency_score:.2f}/100; inspect the mixed themes.",
            )

        cards_by_id = {card.restaurant_id: card for card in cards}
        for category in CATEGORY_ORDER:
            assignment = assignments.get(category)
            if assignment is None:
                continue
            restaurant_id, reason = assignment
            cards_by_id[restaurant_id].categories.append(
                RecommendationBadge(
                    category=category,
                    label=CATEGORY_LABELS[category],
                    reason=reason,
                )
            )
        return [category for category in CATEGORY_ORDER if category in assignments]
